File: carcassonne-master/src/road_side_decorator.cc
#include "road_side_decorator.h"
#include "terrain/road_segment.h"
#include "terrain/field_segment.h"

namespace carcassonne
{

RoadSideDecorator::RoadSideDecorator()
{
}

RoadSideDecorator::RoadSideDecorator(Side* side) : 
	TripleTerrainSideDecorator(side)
{
	segment1_.reset(new FieldSegment);
	middle_.reset(new RoadSegment);
	segment2_.reset(new FieldSegment);
	
}

RoadSideDecorator::RoadSideDecorator(const RoadSideDecorator& src) 
	: TripleTerrainSideDecorator(src)
{
	CopyFrom(src);	
}

RoadSideDecorator& RoadSideDecorator::
	operator=(const RoadSideDecorator& rhs)
{
	if(this == &rhs) {
		return (*this);
	}
	
	CopyFrom(rhs);
	
	return (*this);
}
 	
void RoadSideDecorator::
	CopyFrom(const RoadSideDecorator& src)
{
}
 
Side* RoadSideDecorator::
	Clone() const
{
	return new RoadSideDecorator(*this);
}

SideDecorator* RoadSideDecorator::Decorate(Side* side)
{
	return new RoadSideDecorator(side);
	
}

std::string RoadSideDecorator::ToString() const
{
	return "road " + TerrainSideDecorator::ToString();
	
}

bool RoadSideDecorator::
	IsPotentialTerrainMatch(const TerrainSideDecorator& 
														inTerrainSideDecorator) const
{
	return inTerrainSideDecorator.IsExistingTerrainMatch(*this);
}
 	
bool RoadSideDecorator:: 
	IsExistingTerrainMatch(const FieldSideDecorator& inFieldSideDecorator) const
{
	return false;
}

bool RoadSideDecorator:: 
	IsExistingTerrainMatch(const RoadSideDecorator& inRoadSideDecorator) const
{
	return true;
}

bool RoadSideDecorator::
	IsExistingTerrainMatch(const CitySideDecorator& inCitySideDecorator) const
{
	return false;
}

RoadSideDecorator::~RoadSideDecorator()
{
}

}

File: js/RoadSideDecorator.js
const TripleTerrainSideDecorator=require( "./TripleTerrainSideDecorator");
const TerrainSideDecorator=require( "./TerrainSideDecorator");
const FieldSegment=require( "./terrain/FieldSegment");
const RoadSegment=require( "./terrain/RoadSegment");
module.exports= class RoadSideDecorator extends TripleTerrainSideDecorator {
    constructor(src) {
        super(src);
        this.segment1_ = new FieldSegment();
        this.middle_ = new RoadSegment();
        this.segment2_ = new FieldSegment();
    }

    Clone() {
        return new RoadSideDecorator(this);
    }

    Decorate(side) {
        return new RoadSideDecorator(side);
    }

    ToString() {
        return "road " + TerrainSideDecorator.ToString();

    }

    IsPotentialTerrainMatch(inTerrainSideDecorator) {
        return inTerrainSideDecorator.IsExistingTerrainMatch(this);
    }


    IsExistingTerrainMatch(inRoadSideDecorator) {
        return inRoadSideDecorator instanceof RoadSideDecorator;
    }


}